staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/vault/client.go

```diff
@@ -0,0 +1,220 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package vault implements envelop encryption provider based on Vault KMS
+package vault
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/vault/api"
+)
+
+// Handle all communication with Vault server.
+type clientWrapper struct {
+	client      *api.Client
+	encryptPath string
+	decryptPath string
+	authPath    string
+
+	// We may update token for api.Client, but there is no sync for api.Client.
+	// Read lock for encrypt/decrypt requests, write lock for login requests which
+	// will update token for api.Client.
+	rwmutex sync.RWMutex
+	version uint
+}
+
+// Initialize a client wrapper for vault kms provider.
+func newClientWrapper(config *EnvelopeConfig) (*clientWrapper, error) {
+	client, err := newVaultClient(config)
+	if err != nil {
+		return nil, err
+	}
+
+	// Vault transit path is configurable. "path", "/path", "path/" and "/path/"
+	// are the same.
+	transit := "transit"
+	if config.TransitPath != "" {
+		transit = strings.Trim(config.TransitPath, "/")
+	}
+
+	// auth path is configurable. "path", "/path", "path/" and "/path/" are the same.
+	auth := "auth/"
+	if config.AuthPath != "" {
+		auth = strings.Trim(config.AuthPath, "/")
+	}
+	wrapper := &clientWrapper{
+		client:      client,
+		encryptPath: "/v1/" + transit + "/encrypt/",
+		decryptPath: "/v1/" + transit + "/decrypt/",
+		authPath:    auth + "/",
+	}
+
+	// Set token for the api.client.
+	if config.Token != "" {
+		client.SetToken(config.Token)
+	} else {
+		err = wrapper.refreshToken(config, wrapper.version)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return wrapper, nil
+}
+
+func newVaultClient(config *EnvelopeConfig) (*api.Client, error) {
+	vaultConfig := api.DefaultConfig()
+	vaultConfig.Address = config.Address
+
+	tlsConfig := &api.TLSConfig{
+		CACert:        config.CACert,
+		ClientCert:    config.ClientCert,
+		ClientKey:     config.ClientKey,
+		TLSServerName: config.TLSServerName,
+	}
+	if err := vaultConfig.ConfigureTLS(tlsConfig); err != nil {
+		return nil, err
+	}
+
+	return api.NewClient(vaultConfig)
+}
+
+// Get token by login and set the value to api.Client.
+func (c *clientWrapper) refreshToken(config *EnvelopeConfig, version uint) error {
+	c.rwmutex.Lock()
```

Comment:
I'm fairly certain there is a race condition here if the token can only be used once before having to be renewed that would cause encrypt/decrypt calls to fail since they only retry once.